ENH: stats.ttest_1samp: add array-API support

mdhaber opened the pull request:
#### Reference issue
gh-20544

#### What does this implement/fix?
This adds array-API support to `scipy.stats.ttest_1samp`.

#### Additional Information
```python3
import numpy as np
import numpy as xp
from scipy import stats
rng = np.random.default_rng(34598234592345)
x = xp.asarray(rng.random((1000, 1000)))
stats.ttest_1samp(x, xp.asarray(0.))

# main (NumPy)
# 6.58 ms ± 299 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# pr (NumPy)
# 6.59 ms ± 135 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# pr (PyTorch)
# 3.68 ms ± 172 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

mdhaber committed:
ENH: stats.ttest_1samp: add array-API support

mdhaber committed:
TST: stats.ttest_1samp: convert tests

lucascolley commented on scipy/stats/_stats_py.py:
```diff
@@ -6786,11 +6787,16 @@ class TtestResult(TtestResultBase):
 
     def __init__(self, statistic, pvalue, df,  # public
-                 alternative, standard_error, estimate):  # private
+                 alternative, standard_error, estimate,
```
```suggestion
                 alternative, standard_error, estimate, # private
```

lucascolley commented on scipy/stats/_stats_py.py:
```diff
@@ -6987,35 +6996,49 @@ def ttest_1samp(a, popmean, axis=0, nan_policy='propagate',
-        t = np.divide(d, denom)[()]
-    prob = _get_pvalue(t, distributions.t(df), alternative)
+        t = xp.divide(d, denom)
+        t = t[()] if t.ndim == 0 else t
```
perhaps worth adding a helper if this is going to keep coming up?

mdhaber replied:
It will keep coming up. I'm not sure why, but I'm conflicted about a helper for now. Maybe because it's part of a bigger issue that I think it's worth reminding us about when we read the code? NumPy reducing functions return a different type of object (scalar instead of array) depending on input dimensionality. If NumPy were to switch to returning 0d arrays after reducing operations on 1d elements, we could follow suit and get rid of `[()]` everywhere.

rgommers commented:
Nice! Would it make sense to post a basic benchmark in PR descriptions, to show things work with similar or improved speed compared to using NumPy? E.g., using CuPy if GPU is supported, and PyTorch CPU tensors if only other CPU implementations are supported?

lucascolley commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3660,18 +3677,22 @@ def test_1samp_ci_1d(self, alternative):
-        assert_equal(res.df, n-1)
+        xp_assert_close(ci.low, xp.asarray(ref[alternative][0]))
+        xp_assert_close(ci.high, xp.asarray(ref[alternative][1]))
+        xp_assert_close(res.df, xp.asarray(n-1))
```
Do we not want `xp_assert_equal`?

lucascolley commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3680,17 +3701,18 @@ def test_1samp_ci_iv(self):
+        xp_test = array_namespace(l)  # torch needs `expand_dims`
+        popmean = xp_test.expand_dims(popmean, axis=axis)
+        res = stats.ttest_1samp(data, popmean, alternative=alternative, axis=axis)
+        np.testing.assert_allclose(res.pvalue, xp.asarray(1-alpha, dtype=xp.float64))
```
```suggestion
        xp_assert_close(res.pvalue, xp.asarray(1-alpha, dtype=xp.float64))
```

mdhaber replied:
Oops. Will do.

lucascolley reviewed (commented):
took a look on my phone, may have missed something but looks pretty good!

lucascolley commented:
I am without a GPU for a while now, and I think all of these PRs should run CuPy through the test suite before going in.

mdhaber commented:
Like `pearsonr`, this one is for CPU only until we have a standard way to convert to NumPy arrays for internal calculations (https://github.com/scipy/scipy/issues/18286#issuecomment-2028756647).

mdhaber committed:
MAINT: stats.ttest_1samp: fixups

mdhaber commented:
Addressed the comments and posted timing comparison at the top: NumPy performance is unchanged but PyTorch is considerably faster. One of the Windows tests stopped like:
```
........................................................................ [ 87%]
..............s...................................................s..... [ 87%]
...................................................
Error: The operation was canceled.
```
last time. Not seeing this in other jobs so let's watch out for it.

rgommers commented:
> NumPy performance is unchanged but PyTorch is considerably faster.

Very nice. Even if PyTorch had not been faster this change has value - but almost 2x faster out of the box is great to see.

lucascolley reviewed (approved):
All LGTM. I'm confident that test coverage hasn't reduced and that there aren't any spooky `stats` changes that have gone over my head. Thanks Matt!
